### DISCUSSION TITLE
How to assign port coordinates to variables

### DISCUSSION THREAD

**Discussion by JinchW** [gdsfactory/gdsfactory#3111] [Q&A] [answered]
Hi,

The ports information display of gdsfactory is the most comprehensive in the Python library I have used.

I checked the manual, but there is no example of assigning port coordinates to variables, so I want to ask if there is any related function?

In general, I want to move a ref_1 to the ordinate or abscissa of port 1 of another ref_2.
(P.S. I found that after moving the ref, the displayed port coordinates seemed to be original, that is, the coordinate information was not updated?)

Here is my code (gdsfactory: 8.6.10, Python: 3.10.0)

```
import gdsfactory as gf
c = gf.Component()
mmi1 =  gf.components.mmi(inputs=1, outputs=4, width_taper=1, length_taper=10, length_mmi=5.5, width_mmi=5, gap_input_tapers=0.25, gap_output_tapers=0.25, cross_section='strip')
mmi2 =  gf.components.mmi(inputs=1, outputs=4, width_taper=1, length_taper=10, length_mmi=5.5, width_mmi=5, gap_input_tapers=0.25, gap_output_tapers=0.25, cross_section='strip')

c.add_ref(mmi1).dmove([20,50])
mmi1.pprint_ports()

# The commented area is the function I want to achieve. 
# Maybe the syntax does not conform to the gdsfactory specification.
# coordinate_tempy = mmi1.ports["o1"].y
# c.add_ref(mmi2).dmovey(coordinate_tempy)

c.show()
c.write_gds("demo.gds")
```
![image](https://github.com/user-attachments/assets/824f228c-99fb-4156-81d7-f55a572bb8aa)

**Comment by sebastian-goeldi** [comment]
You need to save the ref to a variable. What you are printing are the ports of the component/cell (gdsfactory terminology vs klayout)
What you want is the coordinates of  reference/instance of a cell (SREF/AREF in gds terminology)

So you need to do:


```python
ref = c.add_ref(mmi1)
ref.dmove([20,50])
ref.pprint_ports()
```

P.S.
You don't need to declare `mmi1` and `mmi2`, they will point to the same component/cell.
So the whole thing would become (`<<` is equivalent to `add_ref`)

```python
mmi = gf.c.mmi(...)

ref1 = c << mmi
ref1.dmove([20,50])
ref1.pprint_ports()
ref2 = c << mmi
ref2.dmove(ref.ports['o1'].center) # here I think you might want to use ref2.connect('o1/2/3/4/5', ref1.ports['o1/2/3/4/5'])
```

Think of cells as blueprints. And references are "I want to build house with blueprint A at coordinates XY". You are printing the ports of the blueprint but you want the ports of the  house at coordinates XY

**Reply by JinchW** [reply]
Thank you for your constructive response, I think I found my answer based on your reply😊.
As I was editing the code, I noticed two issues that I'm not sure are due to my coding style.
1. When I try to run `ref.pprint_ports()`, `'Instance' object has no attribute 'pprint_ports'` prompt will appear. If I don't `add_ref()` the component, like `mmi = gf.c.mmi(...)`, `pprint_ports()` will work.
2. The magnitude of the coordinates obtained from `ref.ports['o1'].center` seems to be exaggerated, in other words, measured in nm instead of μm.
Here is the sample code:
```
c = gf.Component()
mmi =  gf.components.mmi(inputs=1, outputs=4, width_taper=1, length_taper=10, length_mmi=5.5, width_mmi=5, gap_input_tapers=0.25, gap_output_tapers=0.25, cross_section='strip')
ref = c << mmi
ref.dmove([20,50])
# ref.pprint_ports()  it dosen't work
# mmi.pprint_ports()  it work
ref2 = c << mmi
# ref2.dmove(ref.ports['o1'].center[0]) 
print(ref.ports['o1'].center)  # The value is （10000, 50000）
c.show()
c.write_gds("demo.gds")
```

**Reply by sebastian-goeldi** [reply]
My bad sorry. I forgot that it's strictly speaking a gdsfactory function only and does not exist on Reference (yet).

You can get it though with

```python
ref1.ports.print()
```

I just saw there is something missing there: This will print the ports in dbu if the port can be properly represented on the integer grid.

You can get around that with

```python
ref1.ports.copy().print(unit="um")
```

@joamatab probably a good idea to add the function to reference as well. I can also expose the unit on the print of kfactory instance.

**Reply by sebastian-goeldi** [reply, marked as answer]
Also, rereading your question, I am pretty sure you want `ref2.connect('o1', ref1, 'o1')`.

This will move and rotate `ref2` in order to make the two ports aligned to each other.

But if you want to purely move then

`ref2.dmove(ref1.ports['o1'].dcenter)`

should do the trick. Note here `dcenter` vs `center`. Center, x, y, etc. will use the units in DBU, the gds database unit. GDS and OASIS use an integer grid as a basis. In most semi-new, like the last 10 years or even more, use 1nm as a DBU. Therefore, 1um == 1000 DBU.
The um are always `[Dd]{name}`. So a `DPolygon` is in um when `Polygon` is in DBU, `dx` is in um `x` in DBU.
That's why you saw a factor 1000x in `center` because I gave you the wrong info, should be `dcenter`.
Reasoning for all this being somewhat unintuitive is thatfor a lot of objects we inherited/use them from KLayout. And we want to keep a somewhat consistent naming scheme.

**Reply by JinchW** [reply]
Yeah, actually I want `ref2.dmove(ref1.ports['o1'].dcenter)`. Your answer solves the initial needs and subsequent problems.
Thank you! And have a good day!😊

**Reply by MarcSchneiderD** [reply]
This moves the center of ref2 to the port 'o1' position of ref1. What do I have to do to move ref2 so that the port 'o2' of ref2 sits on port 'o1' of ref1?

**Reply by sebastian-goeldi** [reply]
You can use the `move(origin, destination)` for that, i.e. `ref2.move(ref2['o2'].center, ref1['o1'].center)`

**Reply by MarcSchneiderD** [reply]
Thank you very much!!!
